packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js

```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
 			return;
 		}
 
-		menuItemsRef.current = {};
+		const itemIndex = {};
+		menuItems.forEach( ( item ) => {
+			if ( ! itemIndex[ item.parent ] ) {
+				itemIndex[ item.parent ] = [];
+			}
+			itemIndex[ item.parent ].push( item );
+		} );
 
-		const innerBlocks = [];
+		menuItemsRef.current = {};
 
-		for ( const menuItem of menuItems ) {
-			const block = createBlockFromMenuItem( menuItem );
-			menuItemsRef.current[ block.clientId ] = menuItem;
-			innerBlocks.push( block );
-		}
+		const createMenuItemBlocks = ( items ) => {
```

Comment:
We're going to want to unit test as much of this logic as possible. It's probably easiest to do that if we move the logic that turns a flat map of menu items into a tree of blocks out of the React hook and into a pure function that is name-exported.

```js
function createBlocksFromMenuItems( menuItems ) {
	...
}
...
const { blocks, menuItemsByClientId } = createBlocksFromMenuItems( menuItems );
menuItemsRef.current = menuItemsByClientId;
setBlocks( [ createBlock( 'core/navigation', {}, blocks ) ] );
```

This is kind of what I had in mind when I wrote `createBlockFromMenuItem`, but I didn't really think it through properly.

Happy to defer this for when we do https://github.com/WordPress/gutenberg/issues/21285.